### PR TITLE
ACCUMULO-4456 Make NotActiveServiceException an RTE

### PR DESCRIPTION
**Pull request by joshelser**
This would naturally be happening anyways because this
exception is not caught by the Thrift processor. This should
bubble up to the client and the client will naturally retry
the call it was going to make, including a re-lookup of
the master location.

### CODE REVIEW

**Comment by joshelser** [comment]
@keith-turner can you take a look?

**Comment by keith-turner** [comment]
@joshelser have you tested this?  Does it cause a retry?  I suspect it will not cause a retry, but not completely sure.   I think the NotActiveServiceException may cause the client to see a TApplicationException which clients do not retry on, because clients treat this as an unexpected server side exception.

**Comment by joshelser** [comment]
> have you tested this? Does it cause a retry?

So, I tried to test this with via SIGSTOP'ing a client application, flipping the active master, and then SIGCONT'ing the application and I was unable to get the client to actually die without this change. Reading the docs, we would have been throwing a RuntimeException anyways (because the checked exception is not capable of being thrown by our thrift RPC methods), so java would have wrapped it in a UndeclaredThrowableException.

> I suspect it will not cause a retry, but not completely sure

Sounds like we'd have to make it a TException then? I went digging in Thrift code last night trying to remember how this all works...

**Comment by joshelser** [comment]
> Reading the docs, we would have been throwing a RuntimeException anyways (because the checked exception is not capable of being thrown by our thrift RPC methods), so java would have wrapped it in a UndeclaredThrowableException.

I don't think this is 100% clear. What the code does at runtime would be equivalent to what this patch changes (this patch is just more clear). The testing I did was unable to force a failure, but I am not convinced if this is because I didn't figure out how to test it good enough.

**Comment by keith-turner** [comment]
> Sounds like we'd have to make it a TException then? 

I don't think this will work.  But this assertion is based on a fuzzy memory of some code in the past doing that and it not working that way.  However, I am not certain about this.

**Comment by keith-turner** [comment]
>The testing I did was unable to force a failure, but I am not convinced if this is because I didn't figure out how to test it good enough.

This seems really hard to test. One way I can think to test is to write some code for a one time test.

 * Start a master where active is always false.  Do this via a one time code modification.
 * Write some code that calls the internal methods to connect to that master and make a thrift call.

**Comment by joshelser** [comment]
> This seems really hard to test. One way I can think to test is to write some code for a one time test.

:) yeah, something like this is what I was trying to avoid but will probably have to do this today to actually check it. Oh well.

**Comment by joshelser** [comment]
Ok, turns out I had a bug :). I was able to reproduce this and it does cause an AccumuloException to be thrown on the client side.

File: core/src/main/java/org/apache/accumulo/core/client/impl/TableOperationsImpl.java [hunk 1]
```diff
@@ -247,10 +252,14 @@ private void executeFateOperation(long opid, FateOperation op, List<ByteBuffer>
       try {
         client = MasterClient.getConnectionWithRetry(context);
         client.executeFateOperation(Tracer.traceInfo(), context.rpcCreds(), opid, op, args, opts, autoCleanUp);
-        break;
+        return;
       } catch (TTransportException tte) {
         log.debug("Failed to call executeFateOperation(), retrying ... ", tte);
         sleepUninterruptibly(100, TimeUnit.MILLISECONDS);
+      } catch (ThriftNotActiveServiceException e) {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/impl/TableOperationsImpl.java]: could almost use the new java convention where catch can list multiple exceptions, except log msgs differ

File: server/base/src/main/java/org/apache/accumulo/server/rpc/NotActiveServiceException.java [hunk 2]
```diff
@@ -19,7 +19,7 @@
 /**
  * An {@link Exception} which denotes that the service which was invoked is not the active instance for that service in the Accumulo cluster.
  */
-public class NotActiveServiceException extends Exception {
+public class NotActiveServiceException extends RuntimeException {
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/rpc/NotActiveServiceException.java]: Is still still used?

Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/rpc/NotActiveServiceException.java]: its already removed

Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/rpc/NotActiveServiceException.java]: Nope, this is nuked in fb1232f

**Comment by keith-turner** [comment]
I think this looks good @joshelser

**Comment by joshelser** [comment]
Thanks for the prompt review, Keith. Will go ahead and squash+merge.

**Pull request closed without merging**